<a href="https://colab.research.google.com/github/qKTPq/2110574-AI4ENG/blob/main/6670020323_redwine_hw_regression.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
!wget -N https://www.dropbox.com/s/1vvgr9k1sjfjshn/winequality_red.csv

--2025-09-18 15:24:00--  https://www.dropbox.com/s/1vvgr9k1sjfjshn/winequality_red.csv
Resolving www.dropbox.com (www.dropbox.com)... 162.125.80.18, 2620:100:6057:18::a27d:d12
Connecting to www.dropbox.com (www.dropbox.com)|162.125.80.18|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://www.dropbox.com/scl/fi/uhmc99xblgvsbpnw0nsog/winequality_red.csv?rlkey=mrgx37uql8qllqr65kpqmlib7 [following]
--2025-09-18 15:24:00--  https://www.dropbox.com/scl/fi/uhmc99xblgvsbpnw0nsog/winequality_red.csv?rlkey=mrgx37uql8qllqr65kpqmlib7
Reusing existing connection to www.dropbox.com:443.
HTTP request sent, awaiting response... 302 Found
Location: https://ucdf058fbfc6d76764806ef92b7c.dl.dropboxusercontent.com/cd/0/inline/CxkHJjldpyUv7hIZOnnKdGGiHWbplN65FZHNa9fFLdXP6AVXkIUTNayivUUsuVYckqh77Mivd_tr8eKIdX8tLW-29glZ7Cjv8g7xQ5hNYejMjU6z24m5qaRMkZ1fovII6cE/file# [following]
--2025-09-18 15:24:01--  https://ucdf058fbfc6d76764806ef92b7c.dl.dropboxusercontent.com/cd/0/inline

# Task
Analyze the "winequality_red.csv" dataset by performing a linear regression analysis to predict the 'quality' variable. The analysis should include checking the linearity assumption, building and evaluating a linear regression model using a train-test split (test_size=0.25, random_state=101), reporting the model equation, checking for out-of-range predictions, and improving the model through post-processing, followed by a re-evaluation of the model's performance.

## Load the data

### Subtask:
Load the `winequality_red.csv` file into a pandas DataFrame.


**Reasoning**:
Import pandas and load the csv file into a dataframe, then display the head and info of the dataframe to verify the loading process.



In [ ]:
import pandas as pd

df = pd.read_csv('winequality_red.csv')

display(df.head())
display(df.info())

fixed acidity;"volatile acidity";"citric acid";"residual sugar";"chlorides";"free sulfur dioxide";"total sulfur dioxide";"density";"pH";"sulphates";"alcohol";"quality"
0   7.4;0.7;0;1.9;0.076;11;34;0.9978;3.51;0.56;9.4;5                                                                                                                     
1   7.8;0.88;0;2.6;0.098;25;67;0.9968;3.2;0.68;9.8;5                                                                                                                     
2  7.8;0.76;0.04;2.3;0.092;15;54;0.997;3.26;0.65;...                                                                                                                     
3  11.2;0.28;0.56;1.9;0.075;17;60;0.998;3.16;0.58...                                                                                                                     
4   7.4;0.7;0;1.9;0.076;11;34;0.9978;3.51;0.56;9.4;5

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1599 entries, 0 to 1598
Data columns (total 1 columns):
 #   Column                                                                                                                                                                   Non-Null Count  Dtype 
---  ------                                                                                                                                                                   --------------  ----- 
 0   fixed acidity;"volatile acidity";"citric acid";"residual sugar";"chlorides";"free sulfur dioxide";"total sulfur dioxide";"density";"pH";"sulphates";"alcohol";"quality"  1599 non-null   object
dtypes: object(1)
memory usage: 12.6+ KB


None

**Reasoning**:
The data is not loaded correctly as it's in a single column. The separator needs to be specified as ';'. Reload the data with the correct separator and display the head and info again.



In [ ]:
df = pd.read_csv('winequality_red.csv', sep=';')

display(df.head())
display(df.info())

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.4              0.70         0.00             1.9      0.076   
1            7.8              0.88         0.00             2.6      0.098   
2            7.8              0.76         0.04             2.3      0.092   
3           11.2              0.28         0.56             1.9      0.075   
4            7.4              0.70         0.00             1.9      0.076   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 11.0                  34.0   0.9978  3.51       0.56   
1                 25.0                  67.0   0.9968  3.20       0.68   
2                 15.0                  54.0   0.9970  3.26       0.65   
3                 17.0                  60.0   0.9980  3.16       0.58   
4                 11.0                  34.0   0.9978  3.51       0.56   

   alcohol  quality  
0      9.4        5  
1      9.8        5  
2      9.8        5  
3      9.8        6  
4      9.4        5

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1599 entries, 0 to 1598
Data columns (total 12 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   fixed acidity         1599 non-null   float64
 1   volatile acidity      1599 non-null   float64
 2   citric acid           1599 non-null   float64
 3   residual sugar        1599 non-null   float64
 4   chlorides             1599 non-null   float64
 5   free sulfur dioxide   1599 non-null   float64
 6   total sulfur dioxide  1599 non-null   float64
 7   density               1599 non-null   float64
 8   pH                    1599 non-null   float64
 9   sulphates             1599 non-null   float64
 10  alcohol               1599 non-null   float64
 11  quality               1599 non-null   int64  
dtypes: float64(11), int64(1)
memory usage: 150.0 KB


None

## Check linearity assumption

### Subtask:
Calculate the correlation between each feature and the target variable (`quality`).


**Reasoning**:
Calculate the correlation matrix and extract the correlations with the 'quality' column.



In [ ]:
correlation_matrix = df.corr()
quality_correlation = correlation_matrix['quality'].sort_values(ascending=False)
display(quality_correlation)

quality                 1.000000
alcohol                 0.476166
sulphates               0.251397
citric acid             0.226373
fixed acidity           0.124052
residual sugar          0.013732
free sulfur dioxide    -0.050656
pH                     -0.057731
chlorides              -0.128907
density                -0.174919
total sulfur dioxide   -0.185100
volatile acidity       -0.390558
Name: quality, dtype: float64

## Prepare data for modeling

### Subtask:
Split the data into features (X) and target (y), and then split these into training and testing sets using `train_test_split` with `test_size=0.25` and `random_state=101`.


**Reasoning**:
Split the data into features (X) and target (y), and then split these into training and testing sets using train_test_split with test_size=0.25 and random_state=101.



In [ ]:
from sklearn.model_selection import train_test_split

X = df.drop('quality', axis=1)
y = df['quality']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=101)

## Create and train the model

### Subtask:
Build a linear regression model using the selected features and train it on the training data.


**Reasoning**:
Build and train a linear regression model using the training data.



In [ ]:
from sklearn.linear_model import LinearRegression

model = LinearRegression()
model.fit(X_train, y_train)

LinearRegression()

## Report the equation

### Subtask:
Display the coefficients and intercept of the trained linear model.


**Reasoning**:
Access and display the coefficients and intercept of the trained linear regression model.



In [ ]:
print("Coefficients:", model.coef_)
print("Intercept:", model.intercept_)

Coefficients: [ 3.86102372e-02 -1.12507423e+00 -2.89192225e-01  1.43323540e-02
 -1.65340833e+00  1.58492449e-03 -3.12292478e-03 -9.86468102e+00
 -4.06402625e-01  8.68168672e-01  2.98092719e-01]
Intercept: 13.74195399740778


## Predict on the test set

### Subtask:
Make predictions on the test set using the trained model.


**Reasoning**:
Use the trained linear regression model to make predictions on the test feature set.



In [ ]:
predictions = model.predict(X_test)
display(predictions)

array([5.03777143, 5.24675945, 5.0357561 , 5.82917956, 6.2080292 ,
       6.56278055, 6.25595908, 6.07847767, 5.85668209, 5.61642832,
       5.79016994, 5.96145096, 6.22897435, 5.24236116, 5.6014323 ,
       5.22570008, 6.39400155, 5.23017475, 5.20178672, 5.50189436,
       5.83221585, 5.63363375, 5.2703711 , 5.83224934, 5.71221636,
       5.15258177, 5.33975518, 5.66889002, 6.53150727, 5.33419282,
       5.32498956, 5.20347248, 5.76743274, 5.39316538, 5.44611191,
       6.52759152, 5.81583572, 6.27459981, 5.1220189 , 5.10898289,
       6.52660994, 5.24089102, 6.06401261, 5.84902951, 6.26846486,
       6.10005073, 5.75586277, 6.5597829 , 4.90347606, 5.58443004,
       5.48985837, 5.96751678, 5.43574212, 5.4843546 , 5.54870858,
       6.04548449, 6.11067875, 6.03177288, 6.60695125, 6.67103806,
       5.07155914, 4.98088055, 5.42767451, 5.52985503, 5.15370232,
       5.63363375, 5.00101006, 5.55534738, 4.97436112, 5.90604435,
       4.96656459, 5.44888229, 6.04374917, 5.73868806, 5.37891

## Evaluate the model

### Subtask:
Calculate and report the RMSE and R-squared score on the test set.


**Reasoning**:
Calculate and print the RMSE and R-squared scores on the test set using the predicted and actual values.



In [ ]:
from sklearn.metrics import mean_squared_error, r2_score
import numpy as np

rmse = np.sqrt(mean_squared_error(y_test, predictions))
r2 = r2_score(y_test, predictions)

print(f"Root Mean Squared Error (RMSE): {rmse}")
print(f"R-squared (R2) Score: {r2}")

Root Mean Squared Error (RMSE): 0.7270348200967859
R-squared (R2) Score: 0.26868025369836257


## Check for out-of-range predictions

### Subtask:
Determine if any predicted values are outside the valid range for the target variable.


**Reasoning**:
Identify the minimum and maximum quality values in the original dataframe and then check if any of the predictions fall outside this range.



In [ ]:
min_quality = df['quality'].min()
max_quality = df['quality'].max()

out_of_range_predictions = predictions[(predictions < min_quality) | (predictions > max_quality)]

if len(out_of_range_predictions) > 0:
    print(f"Out-of-range predictions were found. Number of out-of-range predictions: {len(out_of_range_predictions)}")
else:
    print("No out-of-range predictions were found.")

No out-of-range predictions were found.


## Improve the model (post-processing)

### Subtask:
Apply post-processing techniques to address out-of-range predictions or further refine the model.


**Reasoning**:
Round the predictions to the nearest integer as wine quality is typically an integer score and display the rounded predictions.



In [ ]:
rounded_predictions = np.round(predictions)
display(rounded_predictions)

array([5., 5., 5., 6., 6., 7., 6., 6., 6., 6., 6., 6., 6., 5., 6., 5., 6.,
       5., 5., 6., 6., 6., 5., 6., 6., 5., 5., 6., 7., 5., 5., 5., 6., 5.,
       5., 7., 6., 6., 5., 5., 7., 5., 6., 6., 6., 6., 6., 7., 5., 6., 5.,
       6., 5., 5., 6., 6., 6., 6., 7., 7., 5., 5., 5., 6., 5., 6., 5., 6.,
       5., 6., 5., 5., 6., 6., 5., 8., 5., 6., 5., 6., 5., 6., 5., 5., 6.,
       5., 5., 6., 5., 6., 7., 6., 5., 6., 6., 5., 5., 6., 5., 6., 5., 6.,
       5., 6., 5., 6., 6., 6., 5., 6., 5., 6., 5., 6., 5., 6., 6., 5., 5.,
       5., 6., 6., 5., 6., 6., 5., 6., 5., 6., 5., 6., 5., 6., 6., 6., 7.,
       5., 6., 6., 6., 5., 6., 5., 5., 6., 7., 6., 6., 6., 5., 6., 6., 5.,
       6., 5., 5., 5., 5., 6., 5., 5., 5., 6., 6., 6., 6., 7., 6., 6., 6.,
       5., 6., 6., 5., 6., 6., 6., 6., 6., 5., 6., 5., 5., 5., 6., 5., 5.,
       6., 6., 5., 6., 6., 6., 6., 6., 6., 5., 6., 6., 6., 5., 6., 5., 7.,
       6., 5., 6., 6., 6., 6., 6., 5., 7., 5., 5., 6., 5., 6., 5., 5., 5.,
       7., 5., 6., 6., 6.

## Re-evaluate the model

### Subtask:
Calculate and report the RMSE and R-squared score on the test set after applying post-processing.


**Reasoning**:
Calculate the RMSE and R-squared score for the rounded predictions and report the results.



In [ ]:
rmse_rounded = np.sqrt(mean_squared_error(y_test, rounded_predictions))
r2_rounded = r2_score(y_test, rounded_predictions)

print(f"Root Mean Squared Error (RMSE) after post-processing: {rmse_rounded}")
print(f"R-squared (R2) Score after post-processing: {r2_rounded}")

Root Mean Squared Error (RMSE) after post-processing: 0.7729812416870153
R-squared (R2) Score after post-processing: 0.17332503199474225


## Summary:

### Data Analysis Key Findings

*   The dataset, using semicolons as separators, was successfully loaded into a pandas DataFrame with 12 columns and no missing values.
*   'Alcohol' showed the strongest positive correlation with 'quality' (0.476), while 'volatile acidity' had the strongest negative correlation (-0.391).
*   A linear regression model was successfully trained on the training data.
*   The initial linear regression model resulted in a Root Mean Squared Error (RMSE) of approximately 0.727 and an R-squared (R2) score of approximately 0.269 on the test set.
*   No out-of-range predictions were found with the initial model; all predictions fell within the actual range of quality values (3 to 8).
*   Rounding the predictions to the nearest integer (post-processing) resulted in an RMSE of approximately 0.773 and an R2 score of approximately 0.173.

### Insights or Next Steps

*   The linear regression model explains only about 27% of the variance in wine quality (based on the initial R2). This suggests that the relationship between the features and quality is likely not purely linear or that other important factors influencing quality are not included in the dataset.
*   Rounding the predictions, while making them more interpretable as integer quality scores, slightly decreased the model's performance metrics (increased RMSE and decreased R2). Further investigation could involve exploring other post-processing techniques or modeling approaches that are better suited for ordinal or discrete target variables like quality.


by **Kantaphan Punnaanan** ID: 6670020323